New network connectivity view

svevang opened the pull request:
<!-- You can remove this first section if you have contributed before -->
### First time contributor checklist
<!-- replace the empty checkboxes [ ] below with checked ones [x] accordingly -->
- [x] I have read the [README](https://github.com/WhisperSystems/Signal-Desktop/blob/master/README.md) and [Contributor Guidelines](https://github.com/WhisperSystems/Signal-Desktop/blob/master/CONTRIBUTING.md)
- [x] I have signed the [Contributor Licence Agreement](https://whispersystems.org/cla/)

### Contributor checklist
<!-- replace the empty checkboxes [ ] below with checked ones [x] accordingly -->
- [x] My contribution is fully baked and ready to be merged as is
- [x] My changes are rebased on the latest master branch
- [x] My commits are in nice logical chunks
- [x] I have followed the [best practices](http://chris.beams.io/posts/git-commit/) in my commit messages
- [x] I have made the choice whether I want the [BitHub reward](https://github.com/WhisperSystems/Signal-Android/wiki/BitHub-Rewards) or not by omitting or adding the word `FREEBIE` in my Git commit messages
- [x] I have tested my contribution on these platforms:
 * Google Chrome 55.0.2883.95 (64-bit)
- [x] My changes pass all the [local tests](https://github.com/WhisperSystems/Signal-Desktop/blob/master/CONTRIBUTING.md#tests) 100%
- [x] I have considered whether my changes need additional [tests](https://github.com/WhisperSystems/Signal-Desktop/blob/master/CONTRIBUTING.md#tests), and in the case they do, I have written them

----------------------------------------

### Description

The PR adds a network status interface, letting the user know if there is a problem with their network connection. If there is an interruption, a small warning panel drops down with a message.

The main sources of network interruptions addressed here:

- [Disconnected networks](https://cloud.githubusercontent.com/assets/329012/21772380/afedc52e-d650-11e6-9d5a-acc24061ff9a.gif)
- Super lossy networks
- [Websocket status and reconnect timeouts](https://cloud.githubusercontent.com/assets/329012/21772381/afee4a6c-d650-11e6-97ac-88b3f4270ba5.gif)

The main point of reference is a new NetworkStatusView. It calculates the overall status of the network and renders messages for the user.

20fd2245af1d3026bad8f35a73c2874311a853ee relocates the instantiation of the inboxView. The goal there is to co-locate the user interface and network/messaging namespaces. Network code can then percolate up events/call methods on the UI classes.

<!--
![lossy-network-reconnecting](https://cloud.githubusercontent.com/assets/329012/21772381/afee4a6c-d650-11e6-97ac-88b3f4270ba5.gif)
![going-offline](https://cloud.githubusercontent.com/assets/329012/21772380/afedc52e-d650-11e6-9d5a-acc24061ff9a.gif)
-->

liliakai commented on js/background.js:
```diff
@@ -277,4 +280,38 @@
         });
     }
+
+    var App = Backbone.Model.extend({
```
Why extend Model here?

svevang replied:
The short is that backbone provides some really nice class semantics, which  e.g. the initialization routine. Maybe a case of YAGNI? My thought was to build out a global 'app' class for mocking and testing.

liliakai commented on js/views/inbox_view.js:
```diff
@@ -156,6 +156,11 @@
 
             var inboxCollection = getInboxCollection();
+
+            inboxCollection.on('messageSendErrors', function(){
```
`s/messageSendError/messageError` to match the event triggered in conversations.js?

liliakai commented on js/views/inbox_view.js:
```diff
@@ -6,34 +6,78 @@
+            this.$el.hide();
+
+            setInterval(this.render.bind(this), 5000);
+            setTimeout(this.finishConnectingGracePeriod.bind(this), 5000);
```
Is this grace period really necessary? This means that if we're still 'CONNECTING' 5 seconds after the window is first opened, we'll see the banner before/until the connection eventually completes or fails?

svevang replied:
The idea is that it generally takes a second for the websocket to connect.  Ostensibly, this is not an error state, so no need to inform the user. When first opening the app, we hide network connectivity errors from the user for a grace period (5 seconds here).

Eliminating this grace period means a "Connecting" message and exclamation appears immediately when booting the app.  Which is actually kind of nice feedback, but maybe too much info for the average user.

liliakai commented on js/background.js:
```diff
@@ -277,4 +280,38 @@
+                try {
+                    if (self.inboxView) { self.inboxView.remove(); }
+                    self.inboxView = new Whisper.InboxView({model: self, window: destWindow});
+                    window.openConversation = function(conversation) {
```
`s/window/destWindow`

References to `window` previously executed in the context of index.js/index.html, must be changed to `destWindow` here. In particular, this breaks opening a conversation from a notification click.

svevang replied:
Thanks! I added a patch that makes sure the function is defined on the destination window.

So this change is a consequence of moving the UI initialization machinery over to the background window.  I see two benefits to doing that. First is that we're grouping all our classes, and functions in the same global scope. So no need to juggle between windows to update view state based on the socket status. The second (hypothetical) is that it may make it easier to insert the signal desktop UI into a tab. Which could be nice if porting to chrome extension.

svevang commented:
updated screenshot of the status view (with svg icon and split message/instructions):

<img width="301" alt="screen shot 2017-02-23 at 12 40 49 pm" src="https://cloud.githubusercontent.com/assets/329012/23293024/390f8a44-fa29-11e6-9fb5-279c6fe4b9d3.png">

liliakai commented:
In 0.36.0
